File: src/types/monitor.ts
/** 单张网卡接口的当前收发速率。 */
export interface NetworkInterface {
  name: string;
  /** 下行（RX）bytes/s；首次采样或计数异常时为 null。 */
  receiveBytesPerSecond: number | null;
  /** 上行（TX）bytes/s；首次采样或计数异常时为 null。 */
  transmitBytesPerSecond: number | null;
}

/** 所选网卡在一个监控快照时刻的双向速率样本。 */
export interface NetworkTrendSample {
  /** 监控快照的 Unix 毫秒时间戳。 */
  timestamp: number;
  /** 下行速率；null 用于展示真实的数据缺口。 */
  receiveBytesPerSecond: number | null;
  /** 上行速率；null 用于展示真实的数据缺口。 */
  transmitBytesPerSecond: number | null;
}

/** 网络采集状态，可区分不可用与成功但没有候选接口。 */
export interface NetworkSnapshot {
  available: boolean;
  interfaces: NetworkInterface[];
}

/** 监控快照，由后端一次性采集并推送，前端只渲染 */
export interface MonitorSnapshot {
  sessionId: string;
  /** Unix 毫秒时间戳 */
  timestamp: number;
  /** 0.0 ~ 100.0；无基线或采集缺失时为 null（未知） */
  cpuUsage: number | null;
  /** 0.0 ~ 100.0；MemTotal/MemAvailable 缺失时为 null（未知） */
  memoryUsage: number | null;
  /** 内存总容量（字节）；后端未上报或采集缺失时为 null */
  memoryTotalBytes: number | null;
  /** 内存已用量（字节）；后端未上报或采集缺失时为 null */
  memoryUsedBytes: number | null;
  /** 0.0 ~ 100.0；df 采集失败时为 null（未知） */
  diskUsage: number | null;
  /** 根分区剩余容量；df 采集失败时为 null */
  diskAvailableBytes: number | null;
  /** 根分区总容量；df 采集失败时为 null */
  diskTotalBytes: number | null;
  network: NetworkSnapshot;
}

/** 长任务状态枚举 */
export enum TaskStatus {
  Pending = 'Pending',
  Running = 'Running',
  Done = 'Done',
  Failed = 'Failed',
}

/** 前端共享投影支持的采样任务类型。 */
export type SamplingTaskType = 'monitor' | 'process';

/** 长任务信息，所有持续任务必须具备 taskId 与状态 */
export interface TaskInfo {
  taskId: string;
  taskType: SamplingTaskType;
  sessionId?: string;
  status: TaskStatus;
  /** Unix 毫秒时间戳 */
  createdAt: number;
  /** 最近的失败详情；仅 Failed 状态可能存在。 */
  error?: AppErrorInfo | null;
}

/** 长任务状态变更事件 payload */
export interface TaskStatusEvent {
  taskId: string;
  taskType: SamplingTaskType;
  sessionId: string;
  status: TaskStatus;
  error?: AppErrorInfo | null;
}
import type { AppErrorInfo } from '@/i18n';

File: src-tauri/src/models/monitor.rs
use serde::{Deserialize, Serialize};

/// 单张网卡接口的当前速率，由相邻累计字节计数计算。
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NetworkInterface {
    /// 网卡接口名称。
    pub name: String,
    /// 接收速率（下行），未知时为 None。
    pub receive_bytes_per_second: Option<u64>,
    /// 发送速率（上行），未知时为 None。
    pub transmit_bytes_per_second: Option<u64>,
}

/// 网络采集结果，available 可区分采集失败和成功但没有候选接口。
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NetworkSnapshot {
    /// /proc/net/dev 是否成功读取并解析。
    pub available: bool,
    /// 按远端返回顺序排列、且不含 lo 的候选网卡。
    pub interfaces: Vec<NetworkInterface>,
}

/// 服务器监控快照，由后端采集后推送给前端渲染
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MonitorSnapshot {
    pub session_id: String,
    /// 采集时间，Unix 毫秒时间戳
    pub timestamp: i64,
    /// CPU 使用率，0.0 ~ 100.0；无基线或采集缺失时为 null（未知）
    pub cpu_usage: Option<f64>,
    /// 内存使用率，0.0 ~ 100.0；MemTotal/MemAvailable 缺失时为 null（未知）
    pub memory_usage: Option<f64>,
    /// 内存总容量，单位字节；MemTotal 缺失/非法时为 null（未知）
    pub memory_total_bytes: Option<u64>,
    /// 内存已用量（MemTotal-MemAvailable），单位字节；任一字段缺失时为 null，
    /// MemAvailable 超出总量时按 0 已用处理，与使用率 clamp 语义一致
    pub memory_used_bytes: Option<u64>,
    /// 磁盘使用率，0.0 ~ 100.0；df 采集失败时为 null（未知）
    pub disk_usage: Option<f64>,
    /// 根分区剩余容量，单位字节；df 采集失败时为 null
    pub disk_available_bytes: Option<u64>,
    /// 根分区总容量，单位字节；df 采集失败时为 null
    pub disk_total_bytes: Option<u64>,
    /// 网络采集状态与全部候选网卡接口速率。
    pub network: NetworkSnapshot,
}

/// 长任务信息，所有持续任务必须可跟踪
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TaskInfo {
    pub task_id: String,
    pub task_type: String,
    /// 关联的会话 ID，部分任务可能不关联会话
    pub session_id: Option<String>,
    pub status: TaskStatus,
    /// 任务创建时间，Unix 毫秒时间戳
    pub created_at: i64,
}

/// 长任务状态枚举
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum TaskStatus {
    Pending,
    Running,
    Done,
    Failed,
}

#[cfg(test)]
#[path = "monitor_test.rs"]
mod tests;
